### PR TITLE
Another go at ensuring search string is cleared when closing interface

### LINKED ISSUE #2300: Image Manipulation PlugIn BUG
Dear Support, please note that after new update of QuickSilver I face a specific bug. Whenevery I make a screenshot via Image Manipulation PlugIn, the Quicksilver window remains on the screenshot (while in previous versions this Quicksilver window was hidden during screenshot snapping). See example attached
screen region 2016-12-31 at 18 57 11

### LINKED ISSUE #2302: 'Type Text' action not working in 1.5.1
The 'Type Text' action appears not to be working anymore in version 1.5.1 (build 401C). OS X 10.11.6 on 2012 rMBP.

Rather than typing the text, the system error bell chimes and Quicksilver dismisses as normal. The 'Paste' and 'Paste as Plain Text' actions work as expected. Quicksilver still has permission under assistive applications. (I assume Type Text works by accessing the virtual keyboard?)

Nothing is reported in the console that seems obviously relevant. However, both invoking and dismissing Quicksilver throws multiple WindowServer errors e.g.:
```
1/1/17 2:01:54.439 PM WindowServer[197]: _CGXRemoveWindowFromWindowMovementGroup: window 0x9e is not attached to window 0x2e7
```
Hex code at the end varies. I'm using the Bezel Interface.

Let me know if I can adjust anything to help you test!

### PR DESCRIPTION
**Pull request by pjrobertson**
The problem is that clearing the search string too early means mnemonics are not saved, so if we move the mnemonic saving code earlier this fixes it.
The reason we want to clear the search string on interface close is a UI thing. See this comment: https://github.com/quicksilver/Quicksilver/pull/2249#issuecomment-236476810
Fixes #2302
Fixes #2300
Related to #2268 #2269 

------

OK guys take a look at this. Basically, I took a look at what @tiennou had done in 2b2cc2dfe4b7a3d4270581527d9ddd130bd8ce6f but realised that that `if (!cont) {}` code is needed but understood why Etienne was doing this - because closing the interface meant the search string was cleared, which meant mnemonics could not be saved (since that was done later on).

Instead, what I did was to make saving the mnemonics happen earlier. Now I'm not sure of the consequences of saving mnemonics *before* an action has been run. Does anybody know?

I've just done some testing, and in fact saving the mnemonics *too late* caused other bugs. 
For example take the an action that accepts an iObject, e.g. `[some folder] ⇥ [make new...] ⇥ [my folder]`

This action returns an object (the newly created folder) and displays it in the interface. The new object is created and returned when the action is executed (see `-[QSInterfaceController executeCommandThreaded] -> -[QSCommand execute] ->` which leads to [here](https://github.com/quicksilver/Quicksilver/blob/master/Quicksilver/Code-QuickStepCore/QSCommand.m#L474) where the interface is cleared), which is *before* the mnemonics were originally saved. This way the mnemonic would never be saved for the original `[some folder]` object, since it's already been cleared for the newly created folder.

Make sense?

In short - we need to be sure that saving the mnemonic before executing the command doesn't cause more harm than good

### CODE REVIEW

**Comment by pjrobertson** [comment]
P.S. do I win an award for the most long-winded commit message?! :D

File: Quicksilver/Code-QuickStepInterface/QSInterfaceController.m [hunk 1]
```diff
@@ -634,10 +632,25 @@ - (void)executeCommand:(id)sender cont:(BOOL)cont encapsulate:(BOOL)encapsulate
...
+	}
+	
 	if (encapsulate) {
 		[self encapsulateCommand];
 		return;
 	}
+	if (!cont) {
+		// this ensures the interface is hidden before an action is run e.g. the 'capture screen region' and 'type text' actions needs this
+		[self hideMainWindowFromExecution:self]; // *** this should only hide if no result comes in like 2 seconds
+	}
```
Comment (tiennou) [inline Quicksilver/Code-QuickStepInterface/QSInterfaceController.m]: If you restore that one, maybe the `if (cont)`-else branch below should go ?

Comment (skurfer) [inline Quicksilver/Code-QuickStepInterface/QSInterfaceController.m]: Agreed. The last else can go.

Comment (pjrobertson) [inline Quicksilver/Code-QuickStepInterface/QSInterfaceController.m]: oh wait yeah, you mean line 660

**Comment by skurfer** [comment]
I’ll go ahead and make the change to the if/else on this branch, and hopefully also add a fix for #2301.

**Comment by pjrobertson** [comment]
sorry, you probably heard the pollution levels in Beijing were through the roof this week (I'm running a startup educating people about air pollution here in Beijing) so I didn't have a second to breath this week. Will make the change now

**Comment by pjrobertson** [comment]
Done

**Comment by skurfer** [comment]
Looks good. Thanks!
